src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_gligen_text_image.py

```diff
@@ -0,0 +1,979 @@
+        else:
+            latents = latents.to(device)
+
+        # scale the initial noise by the standard deviation required by the scheduler
+        latents = latents * self.scheduler.init_noise_sigma
+        return latents
+
+    def enable_fuser(self, enabled=True):
+        for module in self.unet.modules():
+            if type(module) is GatedSelfAttentionDense:
+                module.enabled = enabled
+
+    def draw_inpaint_mask_from_boxes(self, boxes, size):
+        inpaint_mask = torch.ones(size[0], size[1])
+        for box in boxes:
+            x0, x1 = box[0] * size[0], box[2] * size[0]
+            y0, y1 = box[1] * size[1], box[3] * size[1]
+            inpaint_mask[int(y0) : int(y1), int(x0) : int(x1)] = 0
+        return inpaint_mask
+
+    def crop(self, im, new_width, new_height):
+        width, height = im.size
+        left = (width - new_width) / 2
+        top = (height - new_height) / 2
+        right = (width + new_width) / 2
+        bottom = (height + new_height) / 2
+        return im.crop((left, top, right, bottom))
+
+    def target_size_center_crop(self, im, new_hw):
+        width, height = im.size
+        if width != height:
+            im = self.crop(im, min(height, width), min(height, width))
+        return im.resize((new_hw, new_hw), PIL.Image.LANCZOS)
+
+    def project(self, x, hidden_size, device):
+        """
+        x (Batch*768) should be the penultimate feature of CLIP (before projection) projection_matrix (768*768) is the
+        CLIP projection matrix, which should be weight.data of Linear layer defined in CLIP (out_dim, in_dim), thus we
+        need to apply transpose below. this function will return the CLIP feature (without normalziation)
+        """
+        projection_matrix = hf_hub_download(repo_id="anhnct/Gligen_Text_Image", filename="projection_matrix")
+        projection_matrix = torch.load(projection_matrix).type(self.image_encoder.dtype).to(device).T
+        if hidden_size != projection_matrix.shape[0]:
+            return x
+        else:
+            return x @ torch.transpose(projection_matrix, 0, 1)
+
+    def complete_mask(self, has_mask, max_objs, device):
+        mask = torch.ones(1, max_objs).type(self.text_encoder.dtype).to(device)
+        if has_mask is None:
+            return mask
+
+        if isinstance(has_mask, int):
+            return mask * has_mask
+        else:
+            for idx, value in enumerate(has_mask):
+                mask[0, idx] = value
+            return mask
+
+    def get_clip_feature(self, input, hidden_size, device, is_image=False):
+        if is_image:
+            if input is None:
+                return None
+            inputs = self.processor(images=[input], return_tensors="pt").to(device)
+            outputs = self.image_encoder(**inputs)
+            feature = outputs.image_embeds
+            feature = self.project(feature, hidden_size, device).squeeze(0)
+            feature = (feature / feature.norm()) * 28.7
```

Comment:
Where does the magic number come from? I think it should be registered as a configuration variable.